### DISCUSSION TITLE
Server starting on the next port if the port 3000 is currently in use

### DISCUSSION THREAD

**Discussion by dynamic2code** [rails/rails#53198] [Ideas]
Hello guys, am new to rails framework and I find it disturbing that the rails  built in server only serves on port 3000. Can I get some help to change it so that if the port is currently in use it just serves on the next port instead of just crushing ?

**Comment by flavorjones** [comment]
@dynamic2code You can choose a different port with the `-p` option, e.g.:

```
$ bin/rails s -p 3001
=> Booting Puma
=> Rails 8.0.0.beta1 application starting in development
=> Run `bin/rails server --help` for more startup options
Puma starting in single mode...
* Puma version: 6.4.3 (ruby 3.3.5-p100) ("The Eagle of Durango")
*  Min threads: 3
*  Max threads: 3
*  Environment: development
*          PID: 1596753
* Listening on http://127.0.0.1:3001
* Listening on http://[::1]:3001
```

Would that work for you?

**Comment by dynamic2code** [comment]
Can it be made to switch autocratically? Like

1.  your run bin/rails server
2. if 3000 is in use
3. it automatically checks if 3001 is free 
4. if free it serves on that port 
5. if not it checks for 3002 and so on...

**Reply by skipkayhil** [reply]
https://github.com/skipkayhil/dotfiles/blob/606ddc7d3ae150709b2342fae2051142a7e391ee/.config/zsh/functions.d/rho#L7

I've wanted this before so I made a zsh function

**Comment by jeromedalbert** [comment]
You could specify port 0:

```
rails server -p 0
```

This will automatically use a random free port.

FYI port 0 is not specific to Rails, it works with any program where you can specify a port ([source](https://www.lifewire.com/port-0-in-tcp-and-udp-818145)).

**Reply by dynamic2code** [reply]
This does exactly what I was looking for . Thank you @jeromedalbert

**Comment by LEstradioto** [comment]
Ive always struggled with this
I created this [simple gist](https://gist.github.com/LEstradioto/82130ac7e44cf7896e05359283b6e5ef), and it just works.
But the port 0 above is really nice!

Why wouldnt Rails built-in this? Does anyone know if it makes DX worse in any case?
